vowpalwabbit/feature_group.h

```diff
@@ -41,6 +41,38 @@ struct feature
 
 static_assert(std::is_trivial<feature>::value, "To be used in v_array feature must be trivial");
 
+namespace VW
+{
+struct namespace_extent
+{
+  namespace_extent() = default;
+
+  namespace_extent(size_t begin_index, size_t end_index, uint64_t hash)
+      : begin_index(begin_index), end_index(end_index), hash(hash)
+  {
+  }
+
+  namespace_extent(size_t begin_index, uint64_t hash) : begin_index(begin_index), hash(hash) {}
+
+  size_t begin_index = 0;
+  size_t end_index = 0;
+  uint64_t hash = 0;
+
+  friend bool operator==(const namespace_extent& lhs, const namespace_extent& rhs)
+  {
+    return lhs.begin_index == rhs.begin_index && lhs.end_index == rhs.end_index && lhs.hash == rhs.hash;
```

Comment:
Does it make sense to start with the hash first since begin_index is likely to be the same in many cases?

Reply:
yeah that sounds reasonable